src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs

```diff
@@ -491,12 +491,45 @@ async Task<PurgeHistoryResult> IDurableOrchestrationClient.PurgeInstanceHistoryA
             return this.DurabilityProvider.GetOrchestrationStateWithPagination(condition, cancellationToken);
         }
 
+        private static EntityQueryResult ConvertToEntityQueryResult(IEnumerable<DurableEntityStatus> entities, string continuationToken)
+        {
+            var entityQueryResult = new EntityQueryResult();
+            entityQueryResult.Entities = entities;
+            entityQueryResult.ContinuationToken = continuationToken;
+
+            return entityQueryResult;
+        }
+
         /// <inheritdoc />
         async Task<EntityQueryResult> IDurableEntityClient.ListEntitiesAsync(EntityQuery query, CancellationToken cancellationToken)
         {
+            if (!query.IncludeDeleted)
+            {
+                query.FetchState = true;
+            }
+
             var condition = new OrchestrationStatusQueryCondition(query);
             var result = await ((IDurableClient)this).ListInstancesAsync(condition, cancellationToken);
             var entityResult = new EntityQueryResult(result);
+
+            if (!query.IncludeDeleted)
+            {
+                var statefulEntities = entityResult.Entities.Where(e => e.State != null);
+                var statefulEntityResult = ConvertToEntityQueryResult(statefulEntities, entityResult.ContinuationToken);
+
+                while (statefulEntities.ToList().Count < query.PageSize && !string.IsNullOrEmpty(entityResult.ContinuationToken))
+                {
+                    query.ContinuationToken = entityResult.ContinuationToken;
+                    condition = new OrchestrationStatusQueryCondition(query);
+                    result = await ((IDurableClient)this).ListInstancesAsync(condition, cancellationToken);
+                    entityResult = new EntityQueryResult(result);
+                    statefulEntities = entityResult.Entities.Where(e => e.State != null);
+
+                    statefulEntityResult = ConvertToEntityQueryResult(statefulEntities, entityResult.ContinuationToken);
+                }
+                return statefulEntityResult;
```

Comment:
It looks like the final count might actually be larger than `query.PageSize`. For example, if I query for 10 entities, get back a list of 10 with one of them deleted, I'll query for another 10 and potentially return 19 entities. I think this behavior is probably fine, but should update the XML documentation for the `PageSize` to indicate that we may actually return more than the requested number of entities under certain conditions.